src/node_i18n.cc

```diff
@@ -64,7 +79,102 @@ bool InitializeICUDirectory(const char* icu_data_path) {
   }
 }
 
+static void ToUnicode(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_GE(args.Length(), 1);
+  CHECK(args[0]->IsString());
+  Utf8Value val(env->isolate(), args[0]);
+  UErrorCode status = U_ZERO_ERROR;
+  uint32_t options = UIDNA_DEFAULT;
+  UIDNA* uidna = uidna_openUTS46(options, &status);
+  UIDNAInfo info = UIDNA_INFO_INITIALIZER;
+
+  MaybeStackBuffer<char> buf;
+  int32_t len = uidna_nameToUnicodeUTF8(uidna,
+                                        *val, val.length(),
+                                        *buf, buf.length(),
+                                        &info,
+                                        &status);
+
+  if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf.AllocateSufficientStorage(len);
+    len = uidna_nameToUnicodeUTF8(uidna,
+                                  *val, val.length(),
+                                  *buf, buf.length(),
+                                  &info,
+                                  &status);
+  }
+
+  if (U_FAILURE(status)) {
+    env->ThrowError("Cannot convert name to unicode");
+    goto clean;
+  }
+
+  args.GetReturnValue().Set(
+    String::NewFromUtf8(env->isolate(),
+    *buf,
+    v8::NewStringType::kNormal,
+    len).ToLocalChecked());
+
+ clean:
+  uidna_close(uidna);
+}
+
+static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_GE(args.Length(), 1);
+  CHECK(args[0]->IsString());
+  Utf8Value val(env->isolate(), args[0]);
+  UErrorCode status = U_ZERO_ERROR;
+  uint32_t options = UIDNA_DEFAULT;
+  UIDNA* uidna = uidna_openUTS46(options, &status);
+  UIDNAInfo info = UIDNA_INFO_INITIALIZER;
+
+  MaybeStackBuffer<char> buf;
+  int32_t len = uidna_nameToASCII_UTF8(uidna,
+                                       *val, val.length(),
+                                       *buf, buf.length(),
+                                       &info,
+                                       &status);
+
+  if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf.AllocateSufficientStorage(len);
+    len = uidna_nameToASCII_UTF8(uidna,
+                                 *val, val.length(),
+                                 *buf, buf.length(),
+                                 &info,
+                                 &status);
+  }
+
+  if (U_FAILURE(status)) {
+    env->ThrowError("Cannot convert name to ASCII");
```

Comment:
Looks like you can test for  `info.errors & UIDNA_ERROR_DOMAIN_NAME_TOO_LONG`, maybe other values.  The `status` itself is likely to be something like out of memory, invalid parameters, etc.  You could `ThrowError`  including `u_errorName(status)` to return  the ICU status code - that would be good. I don't know if `ThrowError` takes a parameter.